From 82307a75239f165315a5b0cc0a4bd6ad0670e17d Mon Sep 17 00:00:00 2001
From: Michael Adler <adlerm@mailbox.org>
Date: Sun, 11 Sep 2022 16:20:52 +0200
Subject: [PATCH 14/16] center first window

This patch centers window if it is a single window opened. When any
other window opens, then they tile normally, and if all other windows
will close, it will center again.

https://dwm.suckless.org/patches/center_first_window/
---
 config.def.h |  7 ++++---
 dwm.c        | 14 +++++++++++++-
 2 files changed, 17 insertions(+), 4 deletions(-)

diff --git a/config.def.h b/config.def.h
index 67e87ed..13a53ed 100644
--- a/config.def.h
+++ b/config.def.h
@@ -30,9 +30,10 @@ static const Rule rules[] = {
 	 *	WM_CLASS(STRING) = instance, class
 	 *	WM_NAME(STRING) = title
 	 */
-	/* class      instance    title       tags mask     isfloating   monitor */
-	{ "Gimp",     NULL,       NULL,       0,            1,           -1 },
-	{ "Firefox",  NULL,       NULL,       1 << 8,       0,           -1 },
+	/* class      	     instance    title    tags mask     isfloating   CenterThisWindow?     monitor */
+	{ "st",              NULL,       NULL,    0,            0,     	     1,		           -1 },
+	{ "Gimp",            NULL,       NULL,    0,            1,           0,                    -1 },
+	{ "Firefox",         NULL,       NULL,    1 << 8,       0,           0,                    -1 },
 };
 
 /* layout(s) */
diff --git a/dwm.c b/dwm.c
index 5d5b579..8eae225 100644
--- a/dwm.c
+++ b/dwm.c
@@ -107,7 +107,7 @@ struct Client {
 	int basew, baseh, incw, inch, maxw, maxh, minw, minh, hintsvalid;
 	int bw, oldbw;
 	unsigned int tags;
-	int isfixed, isfloating, isurgent, neverfocus, oldstate, isfullscreen;
+	int isfixed, isfloating, isurgent, neverfocus, oldstate, isfullscreen, CenterThisWindow;
 	Client *next;
 	Client *snext;
 	Monitor *mon;
@@ -156,6 +156,7 @@ typedef struct {
 	const char *title;
 	unsigned int tags;
 	int isfloating;
+	int CenterThisWindow;
 	int monitor;
 } Rule;
 
@@ -345,6 +346,7 @@ applyrules(Client *c)
 
 	/* rule matching */
 	c->isfloating = 0;
+                c->CenterThisWindow = 0;
 	c->tags = 0;
 	XGetClassHint(dpy, c->win, &ch);
 	class    = ch.res_class ? ch.res_class : broken;
@@ -357,6 +359,7 @@ applyrules(Client *c)
 		&& (!r->instance || strstr(instance, r->instance)))
 		{
 			c->isfloating = r->isfloating;
+			c->CenterThisWindow = r->CenterThisWindow;
 			c->tags |= r->tags;
 			for (m = mons; m && m->num != r->monitor; m = m->next);
 			if (m)
@@ -1984,6 +1987,15 @@ tile(Monitor *m)
 			if (ty + HEIGHT(c) < m->wh)
 				ty += HEIGHT(c);
 		}
+
+	if (n == 1 && selmon->sel->CenterThisWindow) {
+		double ratio = 0.80, mw = selmon->mw, mh = selmon->mh;
+		resizeclient(selmon->sel,
+			(mw - mw * ratio) / 2,
+			(mh - mh * ratio) / 2,
+			mw * ratio,
+			mh * ratio);
+	}
 }
 
 void
-- 
2.38.1

